File: spark/spark_python/dataframe_nested.py
from typing import Dict, Any

import pandas as pd
from pyspark import SparkContext
from pyspark.sql import SQLContext, DataFrame, Row, functions as F
from pyspark.sql.types import StructField, StructType, IntegerType, DoubleType, StringType, ArrayType, MapType, BooleanType

import class_utils


def show_all(df: DataFrame):
    df.printSchema()
    df.show()


def multiple_country_mean_city_density(sc: SparkContext):
    sql_sc = SQLContext(sc)

    city_schema = StructType([
        StructField('name', StringType()),
        StructField('country', StringType()),
        StructField('area', DoubleType()),
        StructField('population', IntegerType())
    ])
    country_schema = StructType([
        StructField('name', StringType()),
        StructField('iso', StringType()),
        StructField('calling_code', StringType()),
        StructField('area', DoubleType()),
        StructField('population', IntegerType()),
        StructField('capitol', city_schema),
        StructField('cities', ArrayType(city_schema))
    ])

    countries = sql_sc.createDataFrame(pd.DataFrame(class_utils.read_countries_full()), country_schema)  # work without schema too

    countries.registerTempTable('countries')

    result = sql_sc.sql('SELECT '
                        'name, '
                        'TRANSFORM(cities, c -> round(c.population / c.area, 2)) AS cities_density_list '
                        'FROM countries')#.registerTempTable('countries_with_cities_density_list')

    query = """aggregate(
        `{col}`,
        CAST(0.0 AS double),
        (acc, x) -> acc + x,
        acc -> round(acc / size(`{col}`), 2)
    ) AS  `avg_{col}`""".format(col="cities_density_list")

    result = result.selectExpr("*", query)

    show_all(result)


def capitol_to_country_density_ratio(sc: SparkContext):
    sql_sc = SQLContext(sc)

    city_schema = StructType([
        StructField('name', StringType()),
        StructField('country', StringType()),
        StructField('area', DoubleType()),
        StructField('population', IntegerType())
    ])
    country_schema = StructType([
        StructField('name', StringType()),
        StructField('iso', StringType()),
        StructField('calling_code', StringType()),
        StructField('area', DoubleType()),
        StructField('population', IntegerType()),
        StructField('capitol', city_schema),
        StructField('cities', ArrayType(city_schema))
    ])

    countries = sql_sc.createDataFrame(pd.DataFrame(class_utils.read_countries_full()), country_schema)  # work without schema too

    countries.registerTempTable('countries')

    sql_sc.sql('SELECT '
               'name AS country_name, '
               'capitol.name AS capitol_name, '
               'round(population / area, 2) AS country_density, '
               'round(capitol.population / capitol.area, 2) AS capitol_density '
               'FROM countries '
               'WHERE capitol IS NOT NULL').registerTempTable('country_capitol')

    result = sql_sc.sql('SELECT *, '
                        'round(capitol_density / country_density, 2) AS density_ratio '
                        'FROM country_capitol')

    show_all(result)

File: spark/spark_scala/src/main/scala/com/pawdon/growth/DataframeNested.scala
package com.pawdon.growth

import org.apache.spark.sql.{SparkSession, DataFrame, functions => F}

object DataframeNested {
  def showAll(df: DataFrame): Unit = {
    df.printSchema()
    df.show()
  }

  def convert(spark: SparkSession): Unit = {
    val data = JsonClassUtils.readCountriesFull.get
    println(data)
    val df = spark.sqlContext.createDataFrame(data)
    showAll(df)
  }

  def multipleCountryMeanCityDensity(spark: SparkSession): Unit = {
    val countries = spark.sqlContext.createDataFrame(JsonClassUtils.readCountriesFull.get)
    countries.createTempView("countries")

    val subResultColumn = "cities_density_list"

    spark.sqlContext.sql(sqlText =
      s"SELECT " +
        s"name, " +
        s"TRANSFORM(cities, c -> round(c.population / c.area, 2)) AS $subResultColumn " +
        s"FROM countries"
    ).createTempView("countries_cities_density")

    val result = spark.sqlContext.sql(sqlText =
      s"SELECT *, " +
        s"aggregate(`$subResultColumn`, " +
        s"CAST(0.0 AS double), " +
        s"(acc, x) -> acc + x," +
        s"acc -> round(acc / size(`$subResultColumn`), 2)" +
        s") AS  `avg_$subResultColumn`" +
        s"FROM countries_cities_density"
    )

    showAll(result)
  }

  def capitolToCountryDensityRatio(spark: SparkSession): Unit = {
    val countries = spark.sqlContext.createDataFrame(JsonClassUtils.readCountriesFull.get)
    countries.createTempView("countries")

    spark.sqlContext.sql(sqlText =
      "SELECT " +
        "name AS country_name, " +
        "capitol.name AS capitol_name, " +
        "round(population / area, 2) AS country_density, " +
        "round(capitol.population / capitol.area, 2) AS capitol_density " +
        "FROM countries " +
        "WHERE capitol IS NOT NULL"
    ).createTempView("country_capitol")

    val result = spark.sqlContext.sql(sqlText =
      "SELECT *, " +
        "round(capitol_density / country_density, 2) AS density_ratio " +
        "FROM country_capitol"
    )

    showAll(result)
  }

}
